File: Backend/src/main/java/com/duybao/SplitGo/Service/CategoryService.java
package com.duybao.SplitGo.Service;

import com.duybao.SplitGo.DTO.Response.ecommerce.CategoryResponse;
import com.duybao.SplitGo.DTO.request.ecommerce.CreateCategoryRequest;
import java.util.List;

public interface CategoryService {
    List<CategoryResponse> getAllCategories();

    List<CategoryResponse> getRootCategories();

    CategoryResponse createCategory(CreateCategoryRequest request);

    CategoryResponse updateCategory(Long categoryId, CreateCategoryRequest request);

    void deleteCategory(Long categoryId);

    // Admin tree management
    List<CategoryResponse> getCategoryTree();

    long getProductCountByCategory(Long categoryId);

    CategoryResponse createCategoryWithParent(CreateCategoryRequest request, Long parentId);

    CategoryResponse updateCategoryFull(Long categoryId, CreateCategoryRequest request);

    void deleteCategoryWithValidation(Long categoryId);
}

File: Frontend/src/services/categoryService.js
import { parseApiResponse, request } from './apiClient'
import { authFetch } from './authService'

export async function getCategories() {
  const response = await request('/categories')
  const payload = await parseApiResponse(response)
  return Array.isArray(payload?.data) ? payload.data : []
}

// ==================== Admin Category Management ====================

export async function getCategoryTree() {
  const response = await authFetch('/categories/tree')
  const payload = await parseApiResponse(response)
  return Array.isArray(payload?.data) ? payload.data : []
}

export async function createCategory(body, parentId = null) {
  const query = parentId ? `?parentId=${parentId}` : ''
  const response = await authFetch(`/categories${query}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const payload = await parseApiResponse(response)
  return payload?.data || null
}

export async function updateCategory(categoryId, body) {
  const response = await authFetch(`/categories/${categoryId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const payload = await parseApiResponse(response)
  return payload?.data || null
}

export async function deleteCategory(categoryId) {
  const response = await authFetch(`/categories/${categoryId}`, {
    method: 'DELETE',
  })
  await parseApiResponse(response)
}

export async function getCategoryProductCount(categoryId) {
  const response = await request(`/categories/${categoryId}/product-count`)
  const payload = await parseApiResponse(response)
  return payload?.data ?? 0
}

